taichi/backends/vulkan/codegen_vulkan.cpp

```diff
@@ -467,89 +679,187 @@ class TaskCodegen : public IRVisitor {
       TI_NOT_IMPLEMENTED;
     }
     const auto dt = stmt->dest->element_type().ptr_removed();
-    std::string func = "atomicAdd";  // GLSL builtin
-    std::string mem = at_buffer(stmt->dest, dt);
+    
+    spirv::Value addr_ptr;
+    if (ptr_to_buffers_.at(stmt->dest) == BuffersEnum::Root) {
+      addr_ptr = ir_->query_value(stmt->dest->raw_name());
+      addr_ptr.flag = spirv::ValueKind::kVariablePtr; // make this value could store/load
+    } else {
+      addr_ptr = at_buffer(stmt->dest, dt);
+    }
+    spirv::Value data = ir_->query_value(stmt->val->raw_name());
+    spirv::Value val;
     if (dt->is_primitive(PrimitiveTypeID::f32)) {
-      // Buffer has to be specified in the fatomicAdd helpers.
-      const std::string buffer_name =
-          buffer_instance_name(ptr_to_buffers_.at(stmt->dest));
-      func = fmt::format("fatomicAdd_{}", buffer_name);
-      mem = vk_data_address_shifter(stmt->dest, dt);
-    } else if (!is_integral(dt)) {
+      spirv::Value func = ir_->float_atomic_add();
+      val = ir_->make_value(spv::OpFunctionCall, ir_->f32_type(), func, addr_ptr, data);
+    } else if (is_integral(dt)) {
+      val = ir_->make_value(spv::OpAtomicIAdd, ir_->get_primitive_type(dt), addr_ptr,
+      ir_->uint_immediate_number(ir_->u32_type(), 1),
+      ir_->uint_immediate_number(ir_->u32_type(), 0),
+      data);
+    } else {
       TI_ERROR("Vulkan only supports 32-bit atomic data types");
     }
-    // const dt stmt = atomicAdd(mem, val);
-    emit("const {} {} = {}({}, {});", opengl_data_type_name(dt),
-         stmt->raw_name(), func, mem, stmt->val->raw_name());
+    ir_->register_value(stmt->raw_name(), val);
```

Comment:
nit: work-group parallel reduction should be possible in all compute shaders, maybe look into that as an potential optimization?

Reply:
We can leave a issue & comment here

Reply:
Yeah, that requires the backend to support Taichi's Thread Local Storage optimization pass.